File: server/ews/ntlm.ts
/**
 * Minimal NTLMv2 handshake using Node's built-in crypto module.
 *
 * Implements just enough of the NTLM protocol to authenticate a single
 * request against an on-premises Exchange server. No external dependencies.
 *
 * References:
 *   - MS-NLMP: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp
 *   - NTLM type messages: https://davenport.sourceforge.net/ntlm.html
 */
import crypto from 'crypto';

export interface NtlmCredentials {
  username: string;
  password: string;
  domain: string;
}

/** Parse "DOMAIN\\user" into { domain, username }. Plain "user" gets empty domain. */
export function parseCredentials(usernameField: string, password: string): NtlmCredentials {
  const parts = usernameField.split('\\');
  if (parts.length === 2) {
    return { domain: parts[0], username: parts[1], password };
  }
  return { domain: '', username: usernameField, password };
}

// ── Type 1 (Negotiate) ──────────────────────────────────────────────────────

const NTLM_SIGNATURE = Buffer.from('NTLMSSP\0', 'ascii');

export function createType1Message(domain: string): string {
  const domainBytes = Buffer.from(domain.toUpperCase(), 'ascii');
  const workstation = Buffer.from('WORKSTATION', 'ascii');

  // Flags: Negotiate Unicode, Negotiate NTLM, Request Target, Negotiate NTLM2Key
  const flags = 0x00000001 | 0x00000200 | 0x00000004 | 0x00080000 | 0x00008000;

  const buf = Buffer.alloc(32 + domainBytes.length + workstation.length);
  NTLM_SIGNATURE.copy(buf, 0);
  buf.writeUInt32LE(1, 8);          // Type 1
  buf.writeUInt32LE(flags, 12);
  // Domain security buffer
  buf.writeUInt16LE(domainBytes.length, 16);
  buf.writeUInt16LE(domainBytes.length, 18);
  buf.writeUInt32LE(32, 20);
  // Workstation security buffer
  buf.writeUInt16LE(workstation.length, 24);
  buf.writeUInt16LE(workstation.length, 26);
  buf.writeUInt32LE(32 + domainBytes.length, 28);

  domainBytes.copy(buf, 32);
  workstation.copy(buf, 32 + domainBytes.length);

  return buf.toString('base64');
}

// ── Type 2 (Challenge) parsing ───────────────────────────────────────────────

export interface Type2Data {
  challenge: Buffer;
  targetInfo: Buffer;
}

export function parseType2Message(base64: string): Type2Data {
  const buf = Buffer.from(base64, 'base64');

  if (buf.toString('ascii', 0, 7) !== 'NTLMSSP') {
    throw new Error('Invalid NTLM Type 2 message');
  }
  if (buf.readUInt32LE(8) !== 2) {
    throw new Error('Not an NTLM Type 2 message');
  }

  const challenge = buf.subarray(24, 32);

  // Target info security buffer
  let targetInfo = Buffer.alloc(0);
  if (buf.length >= 48) {
    const tiLen = buf.readUInt16LE(40);
    const tiOff = buf.readUInt32LE(44);
    if (tiOff + tiLen <= buf.length) {
      targetInfo = buf.subarray(tiOff, tiOff + tiLen);
    }
  }

  return { challenge, targetInfo };
}

// ── Type 3 (Authenticate) ───────────────────────────────────────────────────

function md4(data: Buffer): Buffer {
  return crypto.createHash('md4').update(data).digest();
}

function hmacMd5(key: Buffer, data: Buffer): Buffer {
  return crypto.createHmac('md5', key).update(data).digest();
}

function ntv2Hash(creds: NtlmCredentials): Buffer {
  const passUnicode = Buffer.from(creds.password, 'utf16le');
  const ntHash = md4(passUnicode);
  const identity = (creds.username.toUpperCase() + creds.domain.toUpperCase());
  const identityUnicode = Buffer.from(identity, 'utf16le');
  return hmacMd5(ntHash, identityUnicode);
}

export function createType3Message(creds: NtlmCredentials, type2: Type2Data): string {
  const ntv2 = ntv2Hash(creds);

  // Build NTLMv2 client challenge (blob)
  const clientNonce = crypto.randomBytes(8);
  const timestamp = ntTimestamp();

  // NTLMv2 blob: 0x01010000 + reserved(4) + timestamp(8) + clientNonce(8) + reserved(4) + targetInfo + reserved(4)
  const blob = Buffer.alloc(28 + type2.targetInfo.length + 4);
  blob.writeUInt32LE(0x00000101, 0);  // blob signature + reserved
  blob.writeUInt32LE(0, 4);           // reserved
  timestamp.copy(blob, 8);
  clientNonce.copy(blob, 16);
  blob.writeUInt32LE(0, 24);          // reserved
  type2.targetInfo.copy(blob, 28);
  blob.writeUInt32LE(0, 28 + type2.targetInfo.length); // trailing reserved

  // NTProofStr = HMAC_MD5(ntv2Hash, serverChallenge + blob)
  const ntProofStr = hmacMd5(ntv2, Buffer.concat([type2.challenge, blob]));
  const ntResponse = Buffer.concat([ntProofStr, blob]);

  // LMv2 response = HMAC_MD5(ntv2Hash, serverChallenge + clientNonce) + clientNonce
  const lmResponse = Buffer.concat([
    hmacMd5(ntv2, Buffer.concat([type2.challenge, clientNonce])),
    clientNonce,
  ]);

  // Session key
  const sessionKey = hmacMd5(ntv2, ntProofStr);

  // Build Type 3 message
  const domainUnicode = Buffer.from(creds.domain.toUpperCase(), 'utf16le');
  const userUnicode = Buffer.from(creds.username, 'utf16le');
  const workstationUnicode = Buffer.from('WORKSTATION', 'utf16le');

  const flags = 0x00000001 | 0x00000200 | 0x00000004 | 0x00080000 | 0x00008000 | 0x00000010;

  // Calculate offsets (fixed header = 72 bytes)
  const headerLen = 72;
  let offset = headerLen;

  const lmOff = offset; offset += lmResponse.length;
  const ntOff = offset; offset += ntResponse.length;
  const domOff = offset; offset += domainUnicode.length;
  const userOff = offset; offset += userUnicode.length;
  const wsOff = offset; offset += workstationUnicode.length;
  const skOff = offset; offset += sessionKey.length;

  const buf = Buffer.alloc(offset);
  NTLM_SIGNATURE.copy(buf, 0);
  buf.writeUInt32LE(3, 8);           // Type 3

  // LM response security buffer
  buf.writeUInt16LE(lmResponse.length, 12);
  buf.writeUInt16LE(lmResponse.length, 14);
  buf.writeUInt32LE(lmOff, 16);

  // NT response security buffer
  buf.writeUInt16LE(ntResponse.length, 20);
  buf.writeUInt16LE(ntResponse.length, 22);
  buf.writeUInt32LE(ntOff, 24);

  // Domain security buffer
  buf.writeUInt16LE(domainUnicode.length, 28);
  buf.writeUInt16LE(domainUnicode.length, 30);
  buf.writeUInt32LE(domOff, 32);

  // User security buffer
  buf.writeUInt16LE(userUnicode.length, 36);
  buf.writeUInt16LE(userUnicode.length, 38);
  buf.writeUInt32LE(userOff, 40);

  // Workstation security buffer
  buf.writeUInt16LE(workstationUnicode.length, 44);
  buf.writeUInt16LE(workstationUnicode.length, 46);
  buf.writeUInt32LE(wsOff, 48);

  // Session key security buffer
  buf.writeUInt16LE(sessionKey.length, 52);
  buf.writeUInt16LE(sessionKey.length, 54);
  buf.writeUInt32LE(skOff, 56);

  // Flags
  buf.writeUInt32LE(flags, 60);

  // Copy payloads
  lmResponse.copy(buf, lmOff);
  ntResponse.copy(buf, ntOff);
  domainUnicode.copy(buf, domOff);
  userUnicode.copy(buf, userOff);
  workstationUnicode.copy(buf, wsOff);
  sessionKey.copy(buf, skOff);

  return buf.toString('base64');
}

/** NTLM timestamp: 100-nanosecond intervals since 1601-01-01. */
function ntTimestamp(): Buffer {
  const epoch = BigInt(Date.now()) * 10000n + 116444736000000000n;
  const buf = Buffer.alloc(8);
  buf.writeBigUInt64LE(epoch);
  return buf;
}

File: internal/ews/ntlm.go
// NTLMv2 handshake (Type 1 / Type 2 / Type 3) implemented on top of Go's
// standard library. Used by the EWS client to authenticate against an
// on-premises Exchange server that has NTLM enabled.
//
// References:
//   - MS-NLMP: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp
//   - NTLM type messages: https://davenport.sourceforge.net/ntlm.html
package ews

import (
	"crypto/hmac"
	"crypto/md5"
	"crypto/rand"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"strings"
	"time"
	"unicode/utf16"

	"golang.org/x/crypto/md4"
)

// Credentials carries the parsed username/domain/password for an NTLM handshake.
type Credentials struct {
	Username string
	Password string
	Domain   string
}

// Type2Data is the subset of a parsed Type 2 challenge that the client needs
// to build a Type 3 response.
type Type2Data struct {
	Challenge  []byte // 8 bytes
	TargetInfo []byte
}

// ntlmSignature is the fixed 8-byte signature at the start of every NTLM message.
var ntlmSignature = []byte("NTLMSSP\x00")

// ParseCredentials splits "DOMAIN\user" into (domain, username). Plain "user"
// yields an empty domain.
func ParseCredentials(userField, password string) Credentials {
	if i := strings.Index(userField, `\`); i >= 0 {
		return Credentials{
			Domain:   userField[:i],
			Username: userField[i+1:],
			Password: password,
		}
	}
	return Credentials{Username: userField, Password: password}
}

// CreateType1Message builds the negotiate message. The domain is uppercased
// ASCII. Returns the raw (pre-base64) bytes.
func CreateType1Message(domain string) []byte {
	domainBytes := []byte(strings.ToUpper(domain))
	workstation := []byte("WORKSTATION")

	// Flags: Negotiate Unicode | NTLM | RequestTarget | NTLM2Key | AlwaysSign
	var flags uint32 = 0x00000001 | 0x00000200 | 0x00000004 | 0x00080000 | 0x00008000

	buf := make([]byte, 32+len(domainBytes)+len(workstation))
	copy(buf, ntlmSignature)
	binary.LittleEndian.PutUint32(buf[8:], 1) // type 1
	binary.LittleEndian.PutUint32(buf[12:], flags)

	// Domain security buffer
	binary.LittleEndian.PutUint16(buf[16:], uint16(len(domainBytes)))
	binary.LittleEndian.PutUint16(buf[18:], uint16(len(domainBytes)))
	binary.LittleEndian.PutUint32(buf[20:], 32)

	// Workstation security buffer
	binary.LittleEndian.PutUint16(buf[24:], uint16(len(workstation)))
	binary.LittleEndian.PutUint16(buf[26:], uint16(len(workstation)))
	binary.LittleEndian.PutUint32(buf[28:], 32+uint32(len(domainBytes)))

	copy(buf[32:], domainBytes)
	copy(buf[32+len(domainBytes):], workstation)
	return buf
}

// ParseType2Message decodes a base64-encoded challenge from WWW-Authenticate.
func ParseType2Message(b64 string) (Type2Data, error) {
	buf, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return Type2Data{}, err
	}
	if len(buf) < 32 {
		return Type2Data{}, errors.New("NTLM Type 2: message too short")
	}
	if string(buf[:7]) != "NTLMSSP" {
		return Type2Data{}, errors.New("NTLM Type 2: bad signature")
	}
	if binary.LittleEndian.Uint32(buf[8:]) != 2 {
		return Type2Data{}, errors.New("NTLM Type 2: not a Type 2 message")
	}

	challenge := append([]byte(nil), buf[24:32]...)

	var targetInfo []byte
	if len(buf) >= 48 {
		tiLen := int(binary.LittleEndian.Uint16(buf[40:]))
		tiOff := int(binary.LittleEndian.Uint32(buf[44:]))
		if tiOff >= 0 && tiOff+tiLen <= len(buf) {
			targetInfo = append([]byte(nil), buf[tiOff:tiOff+tiLen]...)
		}
	}
	return Type2Data{Challenge: challenge, TargetInfo: targetInfo}, nil
}

// CreateType3Message produces the authenticate message.
func CreateType3Message(creds Credentials, t2 Type2Data) []byte {
	ntv2 := ntv2Hash(creds)

	clientNonce := make([]byte, 8)
	_, _ = rand.Read(clientNonce)
	ts := ntTimestamp()

	// NTLMv2 blob
	blob := make([]byte, 28+len(t2.TargetInfo)+4)
	binary.LittleEndian.PutUint32(blob[0:], 0x00000101) // blob signature + reserved
	binary.LittleEndian.PutUint32(blob[4:], 0)          // reserved
	copy(blob[8:], ts)
	copy(blob[16:], clientNonce)
	binary.LittleEndian.PutUint32(blob[24:], 0) // reserved
	copy(blob[28:], t2.TargetInfo)
	// trailing reserved (last 4 bytes already zero)

	// NTProofStr = HMAC_MD5(ntv2, challenge || blob)
	ntProof := hmacMD5(ntv2, concat(t2.Challenge, blob))
	ntResponse := concat(ntProof, blob)

	// LMv2 response = HMAC_MD5(ntv2, challenge || clientNonce) || clientNonce
	lmResponse := concat(hmacMD5(ntv2, concat(t2.Challenge, clientNonce)), clientNonce)

	sessionKey := hmacMD5(ntv2, ntProof)

	domainUnicode := utf16LE(strings.ToUpper(creds.Domain))
	userUnicode := utf16LE(creds.Username)
	workstationUnicode := utf16LE("WORKSTATION")

	var flags uint32 = 0x00000001 | 0x00000200 | 0x00000004 | 0x00080000 | 0x00008000 | 0x00000010

	headerLen := 72
	offset := headerLen

	lmOff := offset
	offset += len(lmResponse)
	ntOff := offset
	offset += len(ntResponse)
	domOff := offset
	offset += len(domainUnicode)
	userOff := offset
	offset += len(userUnicode)
	wsOff := offset
	offset += len(workstationUnicode)
	skOff := offset
	offset += len(sessionKey)

	buf := make([]byte, offset)
	copy(buf, ntlmSignature)
	binary.LittleEndian.PutUint32(buf[8:], 3) // type 3

	writeSecBuf(buf[12:], len(lmResponse), lmOff)
	writeSecBuf(buf[20:], len(ntResponse), ntOff)
	writeSecBuf(buf[28:], len(domainUnicode), domOff)
	writeSecBuf(buf[36:], len(userUnicode), userOff)
	writeSecBuf(buf[44:], len(workstationUnicode), wsOff)
	writeSecBuf(buf[52:], len(sessionKey), skOff)

	binary.LittleEndian.PutUint32(buf[60:], flags)

	copy(buf[lmOff:], lmResponse)
	copy(buf[ntOff:], ntResponse)
	copy(buf[domOff:], domainUnicode)
	copy(buf[userOff:], userUnicode)
	copy(buf[wsOff:], workstationUnicode)
	copy(buf[skOff:], sessionKey)

	return buf
}

// --- helpers ---

func ntv2Hash(c Credentials) []byte {
	pwd := utf16LE(c.Password)
	h := md4.New()
	h.Write(pwd)
	ntHash := h.Sum(nil)

	identity := utf16LE(strings.ToUpper(c.Username) + strings.ToUpper(c.Domain))
	return hmacMD5(ntHash, identity)
}

func hmacMD5(key, data []byte) []byte {
	m := hmac.New(md5.New, key)
	m.Write(data)
	return m.Sum(nil)
}

// utf16LE encodes a Go string as UTF-16 little-endian bytes (no BOM).
func utf16LE(s string) []byte {
	runes := utf16.Encode([]rune(s))
	out := make([]byte, len(runes)*2)
	for i, r := range runes {
		binary.LittleEndian.PutUint16(out[i*2:], r)
	}
	return out
}

// ntTimestamp returns a little-endian 8-byte count of 100-nanosecond intervals
// since 1601-01-01 (the Windows epoch).
func ntTimestamp() []byte {
	// Unix epoch → Windows epoch offset = 11644473600 seconds.
	ns100 := uint64(time.Now().UnixNano()/100) + 116444736000000000
	buf := make([]byte, 8)
	binary.LittleEndian.PutUint64(buf, ns100)
	return buf
}

func writeSecBuf(dst []byte, length, offset int) {
	binary.LittleEndian.PutUint16(dst[0:], uint16(length))
	binary.LittleEndian.PutUint16(dst[2:], uint16(length))
	binary.LittleEndian.PutUint32(dst[4:], uint32(offset))
}

func concat(parts ...[]byte) []byte {
	n := 0
	for _, p := range parts {
		n += len(p)
	}
	out := make([]byte, 0, n)
	for _, p := range parts {
		out = append(out, p...)
	}
	return out
}
